rewriter/Prop.cc

```diff
@@ -337,7 +351,13 @@ vector<ast::ExpressionPtr> processProp(core::MutableContext ctx, PropInfo &ret,
 
     // Compute the setter
     if (!ret.isImmutable) {
-        auto setType = ASTUtil::dupType(ret.type);
+        ast::ExpressionPtr setType;
+        // When we have raise_on_nil_write, the setType should be the non-nilable underlying type
+        if (ret.raiseOnNilWrite && isTNilable(ret.type)) {
+            setType = ASTUtil::dupType(getUnderlyingType(ret.type));
```

Comment:
Could you add a test case that shows that this won't detect `T.any(NilClass, <other-type>)`? That's the interpretation of `T.nilable(<other-type>)`, but this implementation is sensitive to the use of `T.nilable`.